File: services/boss/internal/skillinstall/boss_skill_test.go
package skillinstall

import (
	"strings"
	"testing"
)

// TestBossSkillReferencesHaveNoOrphans asserts that every references/*.md the
// boss core SHIPS is reachable: it must be named by a row of the SKILL.md index
// table, which after BOS-637 is the only route an agent has to a command's
// syntax. An unreferenced reference file is dead weight the payload carries into
// every repo the published core installs into, and — worse — a command group
// silently missing from the index reads as a command group that does not exist.
//
// Only this direction is implemented here. The opposite one (SKILL.md names a
// references/<file>.md the payload does not ship) is already gated across both
// shipped payloads by TestPublishedCoresShipTheReferencesTheyName in
// skills_manifest_test.go; re-implementing it here would add a second copy of an
// invariant that already has a non-vacuity proof.
//
// The payload is read through SkillsFS rather than off the filesystem so this
// gates the bytes that actually get embedded into the binary and extracted into
// a user's skill directory.
func TestBossSkillReferencesHaveNoOrphans(t *testing.T) {
	skill := readEmbeddedBossSkill(t)

	entries, err := SkillsFS.ReadDir("skills/boss/references")
	if err != nil {
		t.Fatalf("read embedded skills/boss/references: %v", err)
	}

	shipped := 0
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		shipped++
		ref := "`references/" + entry.Name() + "`"
		if !strings.Contains(skill, ref) {
			t.Errorf("skills/boss/references/%s is an orphan: SKILL.md's index table has no %s row, "+
				"so nothing routes an agent to it — add the row or delete the file", entry.Name(), ref)
		}
	}
	if shipped == 0 {
		t.Fatalf("skills/boss/references ships no .md files — this test would be vacuous")
	}
}

// TestBossSkillDocumentsBroadcasts pins the broadcast documentation to the skill payload agents
// actually read. Both surfaces exist in code either way, so a refactor that drops this section
// would not fail any build — it would silently un-document the primitive, leaving agents unable
// to discover it.
//
// The rule assertions are short LITERAL anchors, not whole sentences: they survive re-punctuation
// and rewrapping of the surrounding prose, but rewording an anchor itself ("body is a secret" ->
// "the body is secret") does fail. That is the intended trade — these five phrases are the rules
// the ticket requires be stated, so a rewrite that loses one should be a deliberate edit to this
// list rather than a silent drift.
func TestBossSkillDocumentsBroadcasts(t *testing.T) {
	skill := readEmbeddedBossSkill(t)

	// The two names the ticket requires: one per surface.
	assertContains(t, skill, "boss broadcast")
	assertContains(t, skill, "send_broadcast")

	prose := bossBroadcastProseSection(t, skill)

	// The six selector dimensions and the AND/OR combinators.
	for _, dimension := range []string{"chat:", "session:", "repo:", "agent:", "account:", "daemon:"} {
		assertContains(t, prose, dimension)
	}
	assertContains(t, prose, "AND")
	assertContains(t, prose, "OR")

	// Subscriptions and their triggers.
	assertContains(t, prose, "boss broadcast subscribe")
	for _, trigger := range []string{"completed", "errored", "settled"} {
		assertContains(t, prose, trigger)
	}

	// The five rules an agent must internalise.
	for _, rule := range []string{
		"signal, not proof",
		"body is a secret",
		"audience is resolved once",
		"wakes target chats",
		"empty selector is an error",
	} {
		assertContains(t, prose, rule)
	}

	// All six MCP tools, so an MCP-aware host learns the whole surface, not just send.
	for _, tool := range []string{
		"send_broadcast",
		"list_broadcasts",
		"delete_broadcast",
		"register_broadcast_subscription",
		"list_broadcast_subscriptions",
		"delete_broadcast_subscription",
	} {
		assertContains(t, prose, tool)
	}
}

// bossBroadcastProseSection returns the hand-written broadcast section. Its heading is
// deliberately NOT "## Broadcasts": that is the generated CLI reference group earlier in the file,
// and a prose section sharing the name would make these assertions pass against generated command
// help rather than the prose they are meant to pin.
//
// The invariant it actually enforces is positional — the prose must sit AFTER the generated
// region's end marker, because `make gen-skill` rewrites everything between the markers and would
// silently discard a section written inside them. Asserting that directly (rather than counting
// headings) keeps the failure message honest if the generated command group is ever renamed.
func bossBroadcastProseSection(t *testing.T, skill string) string {
	t.Helper()

	const heading = "\n## Broadcasting to other chats\n"
	const endMarker = "<!-- END GENERATED -->"

	generatedEnd := strings.Index(skill, endMarker)
	if generatedEnd == -1 {
		t.Fatalf("marker %q not found in the boss skill", endMarker)
	}
	start := strings.LastIndex(skill, heading)
	if start == -1 {
		t.Fatalf("heading %q not found in the boss skill", strings.TrimSpace(heading))
	}
	if start < generatedEnd {
		t.Fatalf("the %q prose section sits inside the generated region (heading at byte %d, region ends at %d): `make gen-skill` would discard it", strings.TrimSpace(heading), start, generatedEnd)
	}

	prose := skill[start+len(heading):]
	if end := strings.Index(prose, "\n## "); end != -1 {
		prose = prose[:end]
	}
	return prose
}

func readEmbeddedBossSkill(t *testing.T) string {
	t.Helper()

	skillBytes, err := SkillsFS.ReadFile("skills/boss/SKILL.md")
	if err != nil {
		t.Fatalf("read embedded boss skill: %v", err)
	}
	return string(skillBytes)
}

File: scripts/boss-skill.test.mjs
// Content/contract test for the boss skill (BOS-637).
//
// The boss skill is a published core: it installs GLOBALLY, so its resident
// SKILL.md is loaded in every repo on the machine. BOS-637 split the generated
// CLI reference out of that resident body — SKILL.md used to inline all ~92
// command sections (48,624 bytes / 1,171 lines) and now carries only the global
// flags, a routing directive and a 16-row index table pointing at
// references/<group>.md, which an agent opens on demand.
//
// This file is the resident byte ratchet plus the structural invariant that split
// creates: the generated region routes, it does not document. The generator's own
// correctness is gated in Go (skillgen's unit tests, TestSkillMatchesGenerated and
// TestGeneratedReferencesRenderEveryExtractedCommand); what is gated here is the
// thing those tests cannot see — that the resident payload stays small, and that a
// future change cannot quietly inline the command reference back into it.
//
// scripts/Makefile globs scripts/*.test.mjs, so `make test-scripts` runs this.
// `make test-smoke` globs only scripts/bs-*-skill.test.mjs and does NOT.
//
// Node built-ins only — cron worktrees are dependency-free.

import { test } from 'node:test'
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'

import { sectionRegion } from './gate-region-lib.mjs'
import { assertDescendingBudget, measureFile } from './size-ratchet-lib.mjs'

const read = (rel) => readFileSync(new URL(rel, import.meta.url), 'utf8')
const abs = (rel) => fileURLToPath(new URL(rel, import.meta.url))

// BOS-1212: the plugin copy is an rsync of this tree (`make copy-skills`), and
// scripts/skill-mirror-generation.test.mjs asserts that generation once for the whole
// payload. Clauses are pinned against the canonical home only.
const BOSS_CANONICAL = '../services/boss/internal/skillinstall/skills/boss'

const CANONICAL = read(`${BOSS_CANONICAL}/SKILL.md`)

test('size ratchet', () => {
  // A DESCENDING BUDGET, not a ceiling and no longer an exact pin. The number was
  // once the committed size rounded up to the next KiB, compared one-sidedly, so a
  // reduction bought nothing — it just became headroom the resident body could
  // regrow into with nothing going red. The exact pin that replaced it fixed the
  // silence and introduced a different problem: it charged the same repin for a
  // deletion as for an addition, so trimming ceremony cost exactly what adding it
  // did. BOS-1208 prices the two directions apart. Shrinking is FREE — the
  // measurement sits further under the budget and nothing here is touched — while
  // a raise costs a recorded `raise.justification` in the same commit. Never raise
  // this casually: a growing SKILL.md erodes the context budget of EVERY session on
  // the machine, because the boss core installs globally.
  //
  // Pre-split this file was 48624 bytes: the whole generated CLI reference was
  // inline. BOS-637 moved it to references/<group>.md behind an index table,
  // leaving the resident body at ~17 KB. If the reference ever creeps back inline
  // this pin is what catches it — regenerating with `make gen-skill` must not be
  // able to grow the resident payload by ~30 KB unnoticed.
  // Seeded at the measured size, so the budget binds on its very first run rather
  // than starting life as headroom. STEP_DOWN is 512 B because this body is under
  // 20 000 bytes; the bodies at or above that get ~1 KiB, so a bigger body is asked for a bigger
  // step. The share is NOT equal across artifacts, and is deliberately not claimed to be:
  // measured, the step runs from 0.83% of the largest budget (bs-plan, 123354 B) to 3.85% of the
  // smallest in the 1 KiB bucket (bs-sweep-tests, 26600 B), so two buckets narrow the spread a
  // single flat number would give without equalising it.
  const RATCHET = 17402 // measured resident body at migration, 2026-09-08
  const STEP_DOWN = 512
  const REVIEW_BY = '2026-12-08'

  // BOS-1212 removed the guarded two-entry BOSS_MIRRORS list. The vacuity guard that
  // stood here protected the pairing between this pin and the mirror byte-identity test
  // below it; both the list and that test are gone, subsumed by the whole-tree comparison
  // in scripts/skill-mirror-generation.test.mjs. There is no list left to shorten.

  assertDescendingBudget({
    budget: RATCHET,
    constFile: 'scripts/boss-skill.test.mjs',
    constName: 'RATCHET',
    label: 'boss resident SKILL.md',
    measured: measureFile(abs(`${BOSS_CANONICAL}/SKILL.md`)),
    path: 'services/boss/internal/skillinstall/skills/boss/SKILL.md',
    raise: {
      // A LITERAL, deliberately not `RATCHET`. Aliasing the budget constant made this
      // value move in lockstep with every raise, so `budget > from` could never be true
      // and the one direction this primitive prices was free — the arm was structurally
      // dead at every migrated call site (BOS-1208 review). Held at the migration-era
      // measurement, any later raise of RATCHET above it reds until a reason is recorded.
      // No `justification` is pre-supplied either: this commit raised nothing, and a
      // stale sentence parked here would satisfy the next raise without anybody having
      // to write a fresh reason for it, which is the same arm dead a second way.
      from: 17402,
    },
    residual:
      'whether the resident body is any GOOD — only that it fits in this many bytes. A ' +
      'rewrite landing under the budget passes, and the references/ files this body routes ' +
      'to are not measured at all, so content moved out of here is invisible to this budget',
    reviewBy: REVIEW_BY,
    stepDown: STEP_DOWN,
  })
})

test('frontmatter identifies the skill', () => {
  assert.match(CANONICAL, /^---\r?\nname: boss\r?\n/, 'frontmatter must declare name: boss')
})

// generatedRegion returns the bytes between the gen-skill markers — the region
// `make gen-skill` rewrites wholesale. Assertions about what the generator emits
// belong here rather than against the whole file, so hand-written prose elsewhere
// in SKILL.md cannot satisfy them by accident.
const generatedRegion = (skill, label) => {
  const begin = skill.indexOf('<!-- BEGIN GENERATED')
  const end = skill.indexOf('<!-- END GENERATED -->')
  assert.ok(begin !== -1, `${label} must carry the BEGIN GENERATED marker`)
  assert.ok(end > begin, `${label} must carry the END GENERATED marker after BEGIN`)
  return skill.slice(begin, end)
}

test('the generated region routes to per-group references instead of inlining them', () => {
  {
    const dir = BOSS_CANONICAL
    const region = generatedRegion(read(`${dir}/SKILL.md`), `${dir}/SKILL.md`)

    // Global flags stay resident: they apply to every command, so deferring them
    // to a reference would cost a file read on every invocation.
    assert.ok(
      region.includes('## Global Flags'),
      `${dir}: generated region must keep ## Global Flags`,
    )

    // The routing directive is the whole point of the index: an agent must open
    // the reference rather than infer syntax from a one-line index row.
    assert.ok(
      region.includes('**Open the matching reference before using a command**'),
      `${dir}: generated region must carry the "open the reference" routing directive`,
    )

    // The index table itself.
    assert.ok(
      region.includes('## Command Groups'),
      `${dir}: generated region must carry the index heading`,
    )
    assert.match(
      region,
      /^\| Reference\s+\| Read\s+it\s+when…\s+\|$/m,
      `${dir}: index table needs its header row`,
    )
    const rows = region.match(/^\| `references\/[a-z][a-z0-9-]*\.md`\s+\| .+\|$/gm) ?? []
    assert.ok(
      rows.length >= 10,
      `${dir}: index table has ${rows.length} reference rows; expected the full group set`,
    )

    // The invariant the split creates: no command documentation is inline.
    // A `### \`boss …\`` heading here means the reference was re-inlined.
    assert.ok(
      !/^### `boss /m.test(region),
      `${dir}: generated region must not inline command sections — they belong in references/<group>.md`,
    )
  }
})

// ZERO_CHANGE_HEADING names the resident "sessions that change nothing"
// section. The slice is level-aware via sectionRegion().
// The slice is the point: asserting the option names anywhere in SKILL.md would
// pass with them scattered across unrelated sections, which is the shape this
// gate exists to reject. Moving any one row out of the section reds the test.
const ZERO_CHANGE_HEADING = '## Sessions that change nothing'

const assertZeroChangePlacement = (skill, label) => {
  // Position, not just presence: `make gen-skill` rewrites everything between
  // the markers wholesale, so hand-written prose placed above END GENERATED is
  // destroyed on the next regeneration.
  const start = skill.indexOf(ZERO_CHANGE_HEADING)
  assert.ok(start !== -1, `${label} must carry the ${ZERO_CHANGE_HEADING} heading`)
  const endGenerated = skill.indexOf('<!-- END GENERATED -->')
  assert.ok(endGenerated !== -1, `${label} must carry the END GENERATED marker`)
  assert.ok(
    start > endGenerated,
    `${label}: the ${ZERO_CHANGE_HEADING} section must sit AFTER <!-- END GENERATED --> (${start} vs ${endGenerated}) or gen-skill will discard it`,
  )
}

test('the resident body tells an agent how to run a session that changes nothing', () => {
  {
    const dir = BOSS_CANONICAL
    const skill = read(`${dir}/SKILL.md`)
    assertZeroChangePlacement(skill, `${dir}/SKILL.md`)
    const section = sectionRegion(skill, ZERO_CHANGE_HEADING, `${dir}/SKILL.md`)

    for (const option of ['quick_chat', 'defer_pr', '--zero-output']) {
      assert.ok(
        section.includes(option),
        `${dir}: the ${ZERO_CHANGE_HEADING} section must name ${option} — an agent that cannot find all three here defaults to a worktree-and-PR session that finalizes blocked behind an empty draft PR`,
      )
    }

    // The table names the portable create_session field spellings, which read
    // the same on every host. Scoped to the hand-written section on purpose:
    // the generated references/*.md legitimately document the equivalent CLI
    // flags, so this bans the flag spellings here rather than payload-wide.
    for (const notAFlag of ['--quick-chat', '--defer-pr']) {
      assert.ok(
        !section.includes(notAFlag),
        `${dir}: the ${ZERO_CHANGE_HEADING} section must not name ${notAFlag} — this table names the create_session field spellings; the CLI flags belong to the generated command reference`,
      )
    }

    assert.ok(
      section.includes('`create_session`'),
      `${dir}: the ${ZERO_CHANGE_HEADING} section must label quick_chat/defer_pr as create_session fields`,
    )
  }
})
